clippy_utils/src/source.rs

```diff
@@ -283,7 +284,11 @@ impl SourceFileRange {
     /// Attempts to get the text from the source file. This can fail if the source text isn't
     /// loaded.
     pub fn as_str(&self) -> Option<&str> {
-        self.sf.src.as_ref().and_then(|x| x.get(self.range.clone()))
+        self.sf
+            .src
+            .as_ref()
+            .or_else(|| self.sf.external_src.get().and_then(|src| src.get_source()))
```

Comment:
Are there cases where we actually want to return a snippet pointing to an external crate? My gut reaction is that would almost always indicate a macro handling bug

Reply:
`unnecessary_cast` uses it to deal with type aliases.

Reply:
I only noticed `unnecessary_cast` because it has a test for that case. Given that this might silently break lints and I don't really want to audit all uses of it right now it's better to leave it as is.